### PR TITLE
Re-factors Material class so that ANY material may become 'discrete'

### PR DESCRIPTION
**Pull request by aeslaughter**
@bwspenc and @sapitts Would prefer that is be possible to have MOOSE to be able to optionally compute a Material class. Thus, the "DiscreteMaterial" is simply is a Material that MOOSE doesn't compute.

This PR implements this idea by:
- Going back to a single `Material` class.
- Adding a parameter to `Material`: `compute` that indicate if MOOSE should or shouldn't call the compute method.

(refs #6533)

### CODE REVIEW

**Comment by permcody** [comment]
As @jwpeterson would put it: "Classic Slaughter"

Is this really just a revert of your previous PR AND an enhancement? That would certainly be easier to review than looking through "+852 −929".

**Comment by dschwen** [comment]
Andrews previous PR actually broke subtle little things about the zero material by removing old bits of code. Reverting it would reintroduce those, however, we have since reworked that (just as a word of caution...)

**Comment by aeslaughter** [comment]
@dschwen Is correct, but as far is implementation is concerned this is nearly identical and doesn't restore any of the previous bad behavior with dual declare (i.e., that warning that needs to be an error is still in place).

File: modules/phase_field/include/materials/KKSXeVacSolidMaterial.h [hunk 1]
```diff
@@ -15,7 +15,7 @@ class KKSXeVacSolidMaterial;
 template<>
 InputParameters validParams<KKSXeVacSolidMaterial>();
 
-class KKSXeVacSolidMaterial : public DerivativeFunctionMaterialBase
+class KKSXeVacSolidMaterial : public DerivativeFunctionMaterial
```
Comment (dschwen) [inline modules/phase_field/include/materials/KKSXeVacSolidMaterial.h]: Did you really just do a search replace `MaterialBase` -> `Material`?! Do you know about the word boundary marker `\b` for regular expressions?

Comment (dschwen) [inline modules/phase_field/include/materials/KKSXeVacSolidMaterial.h]: This should break so many things!

**Comment by permcody** [comment]
image

**Comment by aeslaughter** [comment]
I have a YAK patch ready.

**Comment by permcody** [comment]
:+1:

**Pull request merged**